### PR TITLE
feat: check auth middleware

### PR DESCRIPTION
**Pull request by lauti7**
This PR:
- adds check_auth middleware to check if we need to fetch the user id from Matrix or if we already have it
- adds function to synapse component to fetch the user id 
- refactor `AppComponents` struct to accept custom components (mainly for testing) 
- add unit tests for check auth middleware

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -74,3 +74,18 @@ else
...
+	@$(RUN_LOCAL_DB)
+	-RUST_LOG=debug cargo test
+	@docker stop social_service_db
+	@rm -rf ./postgres_data
+	@mv ./postgres_data_2 ./postgres_data
+else
+	@$(RUN_LOCAL_DB)
+	-RUST_LOG=debug cargo test
+	@make destroydb
+endif
```
Comment (jmoguilevsky) [inline Makefile:91]:
```suggestion
endif

```

**Review by jmoguilevsky — APPROVED** [review]

**Comment by jmoguilevsky** [comment]
Fix the clippy comments and its ready to go

File: src/components/app.rs [hunk 2]
```diff
@@ -8,47 +9,127 @@ use super::{
...
         let redis = Redis::new_and_run(&config.redis);
 
         // TODO: Should we refactor HealthComponent to avoid cloning structs?
-        health.register_component(Box::new(db.clone()), "database".to_string());
-        health.register_component(Box::new(redis.clone()), "redis".to_string());
+        let health = AppComponents::init_health_component(db.clone(), redis.clone());
+
         let users_cache_instance =
-            users_cache::UsersCacheComponent::new(redis, config.cache_hashing_key.clone());
+            AppComponents::init_users_cache(redis, config.cache_hashing_key.clone());
```
Comment (moliva) [inline src/components/app.rs:59]: nit: Could we replace the type name here as follows?
```suggestion
            Self::init_users_cache(redis, config.cache_hashing_key.clone());
```

File: src/components/app.rs [hunk 3]
```diff
@@ -8,47 +9,127 @@ use super::{
...
 
+        // Initialize components
+        let synapse = AppComponents::init_synapse_component(config.synapse.url.clone());
+        let db = AppComponents::init_db_component(&config.db).await;
         let redis = Redis::new_and_run(&config.redis);
 
         // TODO: Should we refactor HealthComponent to avoid cloning structs?
-        health.register_component(Box::new(db.clone()), "database".to_string());
-        health.register_component(Box::new(redis.clone()), "redis".to_string());
+        let health = AppComponents::init_health_component(db.clone(), redis.clone());
```
Comment (moliva) [inline src/components/app.rs:56]: nit: Could we replace the type name here as follows?
```suggestion
        let health = Self::init_health_component(db.clone(), redis.clone());
```

File: src/components/app.rs [hunk 4]
```diff
@@ -8,47 +9,127 @@ use super::{
...
-        if let Err(err) = db.run().await {
-            log::debug!("Error on running the DB: {:?}", err);
-            panic!("Unable to run the DB")
+    async fn default(config: Config) -> Self {
+        if env_logger::try_init().is_err() {
+            log::debug!("Logger already init")
         }
 
+        // Initialize components
+        let synapse = AppComponents::init_synapse_component(config.synapse.url.clone());
```
Comment (moliva) [inline src/components/app.rs:51]: nit: Could we replace the type name here as follows? Same applies for other comments and wherever the same enclosed type is referred.
```suggestion
        let synapse = Self::init_synapse_component(config.synapse.url.clone());
```

File: src/components/app.rs [hunk 5]
```diff
@@ -8,47 +9,127 @@ use super::{
...
+    pub async fn new(
+        custom_config: Option<Config>,
+        custom_components: Option<CustomComponents>,
+    ) -> Self {
         let config = custom_config
             .unwrap_or_else(|| Config::new().expect("Couldn't read the configuration"));
+        if custom_components.is_none() {
+            AppComponents::default(config).await
+        } else {
+            // For testing propouses mainly
```
Comment (moliva) [inline src/components/app.rs:39]: Spelling for US English:
```suggestion
            // For testing proposes mainly
```

File: src/middlewares/check_auth.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        }
+
+        let token = token.to_string();
+
+        let components = request.app_data::<Data<AppComponents>>().unwrap().clone();
+        let svc = self.service.clone();
+
+        Box::pin(async move {
+            let user_id = {
+                let mut user_cache = components.users_cache.lock().unwrap();
```
Comment (moliva) [inline src/middlewares/check_auth.rs:109]: There's a clippy warning here about the MutexGuard through awaits and suggesting to use an async-aware Mutex instead. Do we know if this affects us?

**Review by moliva — COMMENTED** [review]
Leaving a question and suggestions.

**Pull request merged**